styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesPodEventTranslator.java

```diff
@@ -131,53 +141,104 @@ public TerminationLogMessage(
     }
   }
 
-  // TODO: fix NPath complexity
   static List<Event> translate(
       WorkflowInstance workflowInstance,
       RunState state,
       Pod pod,
       Stats stats) {
 
     final List<Event> generatedEvents = Lists.newArrayList();
-    final Optional<Event> hasError = isInErrorState(workflowInstance, pod);
 
+    final Optional<ContainerStatus> mainContainerStatusOpt = getMainContainerStatus(pod);
+
+    final Optional<Event> hasError = isInErrorState(workflowInstance, pod, mainContainerStatusOpt);
     if (hasError.isPresent()) {
+      handleError(state, generatedEvents, hasError.get());
+      return generatedEvents;
+    }
+
+    final StartedAndExited status = getStartedAndExited(pod, mainContainerStatusOpt);
+
+    handleStarted(workflowInstance, state, generatedEvents, status.started);
+    handleExited(workflowInstance, state, pod, mainContainerStatusOpt, stats, generatedEvents,
+        status.exited);
+
+    return generatedEvents;
+  }
+
+  private static void handleExited(WorkflowInstance workflowInstance, RunState state,
+                                   Pod pod, Optional<ContainerStatus> mainContainerStatusOpt,
+                                   Stats stats,
+                                   List<Event> generatedEvents, boolean exited) {
+    if (exited) {
       switch (state.state()) {
         case PREPARE:
         case SUBMITTED:
+          generatedEvents.add(Event.started(workflowInstance));
+          // intentional fall-through
+
         case RUNNING:
-          generatedEvents.add(hasError.get());
+          final Optional<Integer> exitCode = mainContainerStatusOpt.flatMap(cs ->
+              getExitCodeIfValid(workflowInstance.toKey(), pod, cs, stats));
+          generatedEvents.add(Event.terminate(workflowInstance, exitCode));
           break;
 
         default:
           // no event
           break;
       }
+    }
+  }
 
-      return generatedEvents;
+  private static void handleStarted(final WorkflowInstance workflowInstance, final RunState state,
+                                    final List<Event> generatedEvents, final boolean started) {
+    if (started) {
+      switch (state.state()) {
+        case PREPARE:
+        case SUBMITTED:
+          generatedEvents.add(Event.started(workflowInstance));
+          break;
+
+        default:
+          // no event
+          break;
+      }
     }
+  }
 
-    final PodStatus status = pod.getStatus();
-    final String phase = status.getPhase();
+  private static void handleError(RunState state, List<Event> generatedEvents, Event event) {
+    switch (state.state()) {
+      case PREPARE:
+      case SUBMITTED:
+      case RUNNING:
+        generatedEvents.add(event);
+        break;
 
-    boolean exited = false;
+      default:
+        // no event
+        break;
+    }
+  }
+
+  private static StartedAndExited getStartedAndExited(Pod pod,
```

Comment:
Break this up into two methods, `podExited` and `podStarted` to avoid the need for `StartedAndExited` and get a simpler `translate` method.